linting: Always use latest versions of linting tools; Upgrade clang-format config.

LinZhihao-723 opened the pull request:
# References
<!-- Any issues or pull requests relevant to this pull request -->

# Description
<!-- Describe what this request will change/fix and provide any details necessary for reviewers -->
The current listing tools are set to specific versions in our local linting tasks and GitHub linting workflow. However, these versions are getting outdated. For example, clang has released `clang-format` version 18.1. Linting tool version updating normally includes more formatting standards and bug fixes. We should set the linting tool versions to the latest to apply these updates. Therefore, this PR updates the following things:
- Ensure `lint-requirements.txt`  to install tools with the latest available versions
- Update src files with the latest linting tools and formatting

In the future PR, we should set the workflow to run periodically for health-checking purposes.

# Validation performed
<!-- What tests and validation you performed on the change -->
- Linting tools worked well using `task` locally in Linux.
- Workflow passed, manually checked the tools installed are the latest versions.

kirkrodrigues commented on components/core/src/clp/ir/LogEventDeserializer.cpp:
```diff
@@ -10,8 +10,8 @@
 template <typename encoded_variable_t>
-auto LogEventDeserializer<encoded_variable_t>::create(ReaderInterface& reader)
-        -> BOOST_OUTCOME_V2_NAMESPACE::std_result<LogEventDeserializer<encoded_variable_t>> {
+auto LogEventDeserializer<encoded_variable_t>::create(ReaderInterface& reader
+) -> BOOST_OUTCOME_V2_NAMESPACE::std_result<LogEventDeserializer<encoded_variable_t>> {
```
Shall we report this as a bug?

kirkrodrigues commented on lint-requirements.txt:
```diff
@@ -1,4 +1,4 @@
-clang-format~=17.0
-ruff~=0.1
-yamllint~=1.33
+black>=*
```
Why `>=*` rather than simply `black`

kirkrodrigues replied:
On second thought, we should specify a minimum version since our configs will be version specific, and that would be a good way to indicate what version we configured them for.

LinZhihao-723 replied:
Shall we set them to the latest versions?

kirkrodrigues replied:
Sgtm

kirkrodrigues reviewed (changes requested):
Nice. Since we're now implicitly upgrading to [clang-format 18](https://releases.llvm.org/18.1.0/tools/clang/docs/ClangFormatStyleOptions.html), can we:

* Ensure we're not using any newly deprecated options.
* Add settings for rules introduced in clang-format 18?

LinZhihao-723 commented:
> Nice. Since we're now implicitly upgrading to [clang-format 18](https://releases.llvm.org/18.1.0/tools/clang/docs/ClangFormatStyleOptions.html), can we:
> 
> * Ensure we're not using any newly deprecated options.
> * Add settings for rules introduced in clang-format 18?

- Removed deprecated options in the latest commit
- Added config to clang-format 18 new features

kirkrodrigues reviewed (approved):
How about:

linting: Always use latest versions of linting tools; Upgrade clang-format config. (#384)

LinZhihao-723 commented:
> How about:
> 
> linting: Always use latest versions of linting tools; Upgrade clang-format config. (#384)

lgtm
